src/huggingface_hub/hf_api.py

```diff
@@ -9182,6 +9158,84 @@ def _prepare_folder_deletions(
+
+        Files not matching the `allow_patterns` (allowlist) and `ignore_patterns` (denylist)
+        constraints are discarded.
+        """
+        headers = self._build_hf_headers(token=token)
+
+        folder_path = Path(folder_path).expanduser().resolve()
+        if not folder_path.is_dir():
+            raise ValueError(f"Provided path: '{folder_path}' is not a directory")
+
+        # List files from folder
+        relpath_to_abspath = {
+            path.relative_to(folder_path).as_posix(): path
+            for path in sorted(folder_path.glob("**/*"))  # sorted to be deterministic
+            if path.is_file()
+        }
+
+        # Filter files
+        # Patterns are applied on the path relative to `folder_path`. `path_in_repo` is prefixed after the filtering.
+        filtered_repo_objects = list(
+            filter_repo_objects(
+                relpath_to_abspath.keys(), allow_patterns=allow_patterns, ignore_patterns=ignore_patterns
+            )
+        )
+
+        prefix = f"{path_in_repo.strip('/')}/" if path_in_repo else ""
+
+        # If updating a README.md file, make sure the metadata format is valid
+        # It's better to fail early than to fail after all the files have been uploaded or hashed.
+        for relpath in filtered_repo_objects:
+            if relpath == "README.md":
+                # Create temporary `CommitOperationAdd` to provide `as_file`
+                addition = CommitOperationAdd(
+                    path_or_fileobj=relpath_to_abspath[relpath],  # absolute path on disk
+                    path_in_repo=prefix + relpath,  # "absolute" path in repo
+                )
+                with addition.as_file() as file:
+                    response = get_session().post(
+                        f"{ENDPOINT}/api/validate-yaml",
+                        json={"content": file.read().decode(), "repoType": repo_type},
+                        headers=headers,
+                    )
+                    # Handle warnings (example: empty metadata)
+                    response_content = response.json()
+                    message = "\n".join(
+                        [f"- {warning.get('message')}" for warning in response_content.get("warnings", [])]
+                    )
+                    if message:
+                        warnings.warn(f"Warnings while validating metadata in README.md:\n{message}")
+
+                    # Raise on errors
+                    try:
+                        hf_raise_for_status(response)
+                    except BadRequestError as e:
+                        errors = response_content.get("errors", [])
+                        message = "\n".join([f"- {error.get('message')}" for error in errors])
+                        raise ValueError(f"Invalid metadata in README.md.\n{message}") from e
+                # Skip other additions after `README.md` has been processed
+                break
```

Comment:
Could you move this logic to a separate method like this?

```py
def _validate_yaml(self, content: str, *, repo_type: Optional[str] = None, token: Union[bool, str, None] = None):
    # takes as input the raw content to validate
    # make the request + print warnings + raise on error
    (...)
```

This would make the code more readable by been self-contained and separated from the `_prepare_upload_folder_additions` logic.

Reply:
Done, would you like a docstring as well?

Reply:
That would be great yes! Thanks!